features/enlinkd/org.opennms.features.enlinkd.topogen/src/main/java/org/opennms/enlinkd/topogen/protocol/Protocol.java

```diff
@@ -0,0 +1,184 @@
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.enlinkd.topogen.protocol;
+
+import java.net.InetAddress;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import org.opennms.enlinkd.topogen.util.InetAddressGenerator;
+import org.opennms.enlinkd.topogen.TopologyGenerator;
+import org.opennms.enlinkd.topogen.TopologyGenerator.Topology;
+import org.opennms.enlinkd.topogen.TopologyPersister;
+import org.opennms.enlinkd.topogen.topology.LinkedPairGenerator;
+import org.opennms.enlinkd.topogen.topology.PairGenerator;
+import org.opennms.enlinkd.topogen.topology.RandomConnectedPairGenerator;
+import org.opennms.enlinkd.topogen.topology.UndirectedPairGenerator;
+import org.opennms.enlinkd.topogen.util.RandomUtil;
+import org.opennms.netmgt.model.OnmsIpInterface;
+import org.opennms.netmgt.model.OnmsNode;
+import org.opennms.netmgt.model.OnmsSnmpInterface;
+import org.opennms.netmgt.model.monitoringLocations.OnmsMonitoringLocation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public abstract class Protocol<Element> {
+
+    private final static Logger LOG = LoggerFactory.getLogger(CdpProtocol.class);
+
+    protected final Topology topology;
+    protected final int amountNodes;
+    protected final int amountLinks;
+    protected final int amountElements;
+    private final int amountSnmpInterfaces;
+    private final int amountIpInterfaces;
+    protected final TopologyPersister persister;
+    protected final RandomUtil random = new RandomUtil();
+
+    @java.beans.ConstructorProperties({"topology", "amountNodes", "amountLinks", "amountElements",
+        "amountSnmpInterfaces", "amountIpInterfaces", "persister"})
+    public Protocol(Topology topology, int amountNodes, int amountLinks, int amountElements,
+        int amountSnmpInterfaces, int amountIpInterfaces, TopologyPersister persister) {
+        this.topology = topology;
+        this.amountNodes = amountNodes;
+        this.amountLinks = amountLinks;
+        this.amountElements = amountElements;
+        this.amountSnmpInterfaces = amountSnmpInterfaces;
+        this.amountIpInterfaces = amountIpInterfaces;
+        this.persister = persister;
+    }
+
+    public void createAndPersistNetwork() throws SQLException {
+        LOG.info("creating {} {} topology with {} {}, {} {}, {} {}, {} {}, {} {}.",
+                this.topology,
+                this.getProtocol(),
+                this.amountNodes, "Nodes" ,
+                this.amountElements, "Elements",
+                this.amountLinks, "Links",
+                this.amountSnmpInterfaces, "SnmpInterfaces",
+                this.amountIpInterfaces, "IpInterfaces");
+        List<OnmsNode> nodes = createNodes(amountNodes);
```

Comment:
This looks weird. Can we not just add the Strings "Nodes", "Elements", "links" etc. instead of dynamically adding them if they are constants anyways? The log line reads much easier in that case.

Reply:
:+1: This is a typical refactor artifact :-( , fixed